app/functoria_app.ml

```diff
@@ -519,23 +518,239 @@ module Make (P: S) = struct
+    >>= fun files ->
+    List.fold_left (fun r p -> r >>= fun () -> Bos.OS.Path.delete p)
+      (R.ok ()) files
+    >>= fun () ->
+    with_current build_dir
+      (fun () ->
+        let target_dir = build_config_dir in
+        (* Import files in the build target *)
+        let target_path = Fpath.(target_dir / file_ml)
+        and target_path_dune = Fpath.(target_dir / "dune.config")
+        in
+        Bos.OS.Dir.create target_dir
+        >>= fun _ ->
+        Bos.OS.Path.symlink
+          ~force:true
+          ~target:(Fpath.(v ".." // config_dir_to_build_dir / file_ml))
+          target_path
+        >>= fun () ->
+        Bos.OS.Path.exists Fpath.(config_dir_to_build_dir / "dune.config")
+        >>= fun res ->
+        (match res with
+        | true -> (
+          Bos.OS.Path.symlink
+            ~force:true
+            ~target:(Fpath.(v ".." // config_dir_to_build_dir / "dune.config"))
+            target_path_dune
+          >>= fun () -> Ok true)
+        | false -> Ok false)
+        (* Generate dune configuration file *)
+        >>= fun has_dune_inc ->
+        let dune_file = Fpath.(target_dir / "dune")
+        and dune_project_file = Fpath.(v "dune-project") in
+        let pkgs = match P.packages with
+          | []   -> ""
+          | pkgs ->
+            let pkgs =
+              List.fold_left (fun acc pkg ->
+                  String.Set.union pkg.ocamlfind acc
+                ) String.Set.empty pkgs
+              |> String.Set.elements
+            in
+            let pkgs = if has_dune_inc then "config_custom"::pkgs else pkgs
+            in
+            String.concat ~sep:" " pkgs
+        in
+        let dune_content_default =
+          "(executable (name config) (modules config) (libraries "^pkgs^"))" in
+        let dune_content =
+          if has_dune_inc
+          then ("(include dune.config)"^dune_content_default)
+          else dune_content_default in
+        let write_dune_file =
+          Bos.OS.File.delete dune_file
+          >>= fun () -> Bos.OS.File.write dune_file dune_content
+        and write_dune_project_file = Bos.OS.File.exists dune_project_file
+        >>= function | false -> Bos.OS.File.write dune_project_file "(lang dune 1.0)"
+                      | true  -> Ok ()
+        in
+        (* Build config.cmxs with dune *)
+        let target_file = Fpath.(v "_build" / "default" // target_dir / file) |> Fpath.to_string in
+        let cmd =
+          Bos.Cmd.(v "dune" % "build" % "--no-print-directory" % "--root" % "." % target_file)
+        in
+        write_dune_project_file >>= fun () -> write_dune_file >>= fun () ->
+        Bos.OS.Cmd.run_out cmd |> Bos.OS.Cmd.out_string >>= fun (out, status) ->
+        match snd status with
+        | `Exited 0 ->  Ok (Fpath.(to_string (target_dir / file)))
+        | `Exited _
+        | `Signaled _ ->
+          Format.fprintf Format.str_formatter "error while executing %a\n%s"
```

Comment:
This is:

```ocaml
let err = Fmt.strf "error while executing %a\n%s" Bos.Cmd.pp cmd out in
Error (`Invalid_config_ml  x)
```

It would even be better to move the creation of error values in separate functions (so it's easier to spot the different failure modes in one go) but this is not necessary to do for the given PR.
and `| `Signaled _ -> err_invalid_config_ml cmd out`

Reply:
I haven't separated functions but I updated it according to your review